### PR TITLE
Explode missiles at end of life

### PR DESCRIPTION
**Pull request by nallath**
It makes larger battles a bit more interesting. I've pitted a bunch of groups against each other and it results in more dynamic fights (because groups will be dispersed). 

The behavior isn't quite perfect, because the AI only tries to avoid the missiles somewhat. This is most notable when a missile is moving towards a ship (In that case it won't realize that flying in front of a nuke is also a **very bad idea™**

It does expose a bit of an issue if there are lots and lots of nukes out there, at which point the AI ships get a bit of paralysis, since they are surrounded by avoid areas. This is something that could be fixed by making a distinction between "You can't fly here, just don't" areas and "Try to avoid this if possible, because it's not good for your survival" areas. 

As usual; I'm looking for some feedback as to what people think of this.

### CODE REVIEW

**Comment by daid** [comment]
> It does expose a bit of an issue if there are lots and lots of nukes out there, at which point the AI ships get a bit of paralysis,

What would you do if you where surrounded by nukes? (Probably shout "Naar links draaien! Naar links draaien!")
Note that the avoidance code isn't perfect, as seen when AI ships fly tough minefields and asteroid fields. So I think the many nukes issue won't really be a huge issue.

Maybe the avoid area should be "leading" the missile by some distance. After all, flying behind a nuke is less of an issue then in front of it.

**Comment by nallath** [comment]
It's mostly speculative how this could be solved (and based on some courses about this that I took 8+ years ago). But it's probably not worth the effort to actually solve the issue (because it's not that big).

Letting the explode area lead a bit might be a good one, but they will still make a bit of mistakes when the missile turns (but I do expect that this will be way less as what is currently going on). 

Another option that I just came up with is to simply increase the size over time, which should also leave more area open to traverse. This also makes a bit of sense, since flying close to the nuke in the start is not that dangerous but it is dangerous at the end.

File: src/spaceObjects/missiles/nuke.h [hunk 1]
```diff
@@ -8,10 +8,15 @@ class Nuke : public MissileWeapon
     constexpr static float blast_range = 1000.0f;
     constexpr static float damage_at_center = 160.0f;
     constexpr static float damage_at_edge = 30.0f;
+    bool avoid_area_added;
 public:
     Nuke();
 
     virtual void hitObject(P<SpaceObject> object);
+    void explode();
+    void lifeEnded();
```
[the remark this answers is not in the document]
Comment (nallath) [reply src/spaceObjects/missiles/nuke.h]: It doesn't have to be, but it's probably clearer if they are both marked as such (both virtual / override)

File: src/spaceObjects/missiles/EMPMissile.h [hunk 2]
```diff
@@ -8,10 +8,14 @@ class EMPMissile : public MissileWeapon
     constexpr static float blast_range = 1000.0f;
     constexpr static float damage_at_center = 160.0f;
     constexpr static float damage_at_edge = 30.0f;
+    bool avoid_area_added;
 public:
     EMPMissile();
     
     virtual void hitObject(P<SpaceObject> object);
+    void explode();
+    void lifeEnded();
```
[the remark this answers is not in the document]
Comment (nallath) [reply src/spaceObjects/missiles/EMPMissile.h]: Yeah, it works just fine. Some compilers might warn about the virtual not being set explicitly.

File: src/spaceObjects/missiles/nuke.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (daid) [reply src/spaceObjects/missiles/nuke.h]: Note, i will most like add -Werror=suggest-override as compile flag at some point. To ensure overrides are explicit.

I think -Wshadows checks for the missing virtual. But that will cause a lot of other warnings as well.

[the remark this answers is not in the document]
Comment (daid) [reply src/spaceObjects/missiles/nuke.h]: Damn, that plan won't work. Box2D doesn't have override markers, and I don't want to modify that. Do see a lot of missing overrides in my own code as well, not surprising as I didn't know about that keyword when I started making EE.

[the remark this answers is not in the document]
Comment (amir-arad) [reply src/spaceObjects/missiles/nuke.h]: I've learned many new C++ features from EE

[the remark this answers is not in the document]
Comment (daid) [reply src/spaceObjects/missiles/nuke.h]: > I've learned many new C++ features from EE

And there are a bunch more that I'm using now in SP2. Variadic templates are one of those things that I discovered afterwards, and solves a bunch of issues. Especially with code duplication in the script engine bindings.

std::chrono is just nice for not needing to depend on platform specific stuff for time keeping (especially std::chrono::steady_clock is nice)

And defining your own iterators allows range for loops: https://en.cppreference.com/w/cpp/language/range-for
Removing the need for this macro: https://github.com/daid/SeriousProton/blob/master/src/P.h#L258

**Comment by nallath** [comment]
Whoops! I completely forgot about this. I will have a look at this this week.

**Comment by nallath** [comment]
@Triqueon As far as I'm concerned, this is ready now!

**Comment by Triqueon** [comment]
Awesome. @daid, any chance we could merge this?

**Comment by Triqueon** [comment]
(and, pretty please, get a new build/release afterwards :-p ? )

**Comment by nallath** [comment]
Bump?

**Comment by Triqueon** [comment]
Bump again @daid ?

**Pull request merged**